package ej222pj_assign2.exercise_3;

import static org.junit.Assert.*;
import ej222pj_assign2.exercise_1.ArrayIntList;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

public class ArrayIntListTest {
	ArrayIntList testList;
	@Before
	public void setUp(){
		testList = Build(10);
	}
	
	@After
	public void tearDown(){
		
	}
	
	@Test
	public void testGetAtIndexOne() {
		int recievedNumberAtIndexOne = testList.get(1);
		assertEquals(1, recievedNumberAtIndexOne);
	}
	
	@Test(expected=IndexOutOfBoundsException.class)
	public void testGetAtOutOfBounds() {
		testList.get(11);
	}
	
	@Test
	public void testAddFiveToList() {
		int listSize = 5;
		ArrayIntList list = Build(listSize);
		assertEquals(listSize, list.size());
	}
	
	@Test
	public void testAddThreeHundredToList() {
		int listSize = 300;
		ArrayIntList list = Build(listSize);
		assertEquals(listSize, list.size());
	}

	@Test
	public void testAddAtIndexOne() {
		int numberToAddAtIndexOne = 10;
		testList.addAt(numberToAddAtIndexOne, 1);
		int addedNumberAtIndexOne = testList.get(1);
		assertEquals(numberToAddAtIndexOne, addedNumberAtIndexOne);
	}
	
	@Test(expected=IndexOutOfBoundsException.class)
	public void testAddAtOutOfBounds() {
		int OutOfBoundsIndex = 11;
		testList.addAt(1, OutOfBoundsIndex);
	}

	@Test
	public void testRemoveAtIndexOne() {
		int currentNumberAtIndexOne = testList.get(1);
		testList.remove(1);
		int numberAfterRemoveAtIndexOne = testList.get(1);
		assertNotSame(currentNumberAtIndexOne, numberAfterRemoveAtIndexOne);
	}
	
	@Test(expected=IndexOutOfBoundsException.class)
	public void testRemoveAtOutOfBounds() {
		testList.remove(11);
	}

	@Test
	public void testIndexOfOne() {
		int indexOfOne = testList.indexOf(1);
		assertEquals(1, indexOfOne);
	}
	
	public ArrayIntList Build(int size) {
		ArrayIntList intList = new ArrayIntList();
		
		for(int i = 0; i < size; i++) {
			intList.add(i);
		}
		
		return intList;
	}

}
